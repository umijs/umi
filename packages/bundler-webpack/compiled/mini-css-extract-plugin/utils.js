"use strict";

const NativeModule = require("module");
const path = require("path");

/** @typedef {import("webpack").Compilation} Compilation */
/** @typedef {import("webpack").Module} Module */
/** @typedef {import("webpack").LoaderContext<any>} LoaderContext */

/**
 * @returns {boolean}
 */
function trueFn() {
  return true;
}

/**
 * @param {Compilation} compilation
 * @param {string | number} id
 * @returns {null | Module}
 */
function findModuleById(compilation, id) {
  const {
    modules,
    chunkGraph
  } = compilation;
  for (const module of modules) {
    const moduleId = typeof chunkGraph !== "undefined" ? chunkGraph.getModuleId(module) : module.id;
    if (moduleId === id) {
      return module;
    }
  }
  return null;
}

/**
 * @param {LoaderContext} loaderContext
 * @param {string | Buffer} code
 * @param {string} filename
 * @returns {object}
 */
function evalModuleCode(loaderContext, code, filename) {
  // @ts-ignore
  const module = new NativeModule(filename, loaderContext);

  // @ts-ignore
  module.paths = NativeModule._nodeModulePaths(loaderContext.context); // eslint-disable-line no-underscore-dangle
  module.filename = filename;
  // @ts-ignore
  module._compile(code, filename); // eslint-disable-line no-underscore-dangle

  return module.exports;
}

/**
 * @param {string} a
 * @param {string} b
 * @returns {0 | 1 | -1}
 */
function compareIds(a, b) {
  if (typeof a !== typeof b) {
    return typeof a < typeof b ? -1 : 1;
  }
  if (a < b) {
    return -1;
  }
  if (a > b) {
    return 1;
  }
  return 0;
}

/**
 * @param {Module} a
 * @param {Module} b
 * @returns {0 | 1 | -1}
 */
function compareModulesByIdentifier(a, b) {
  return compareIds(a.identifier(), b.identifier());
}
const MODULE_TYPE = "css/mini-extract";
const AUTO_PUBLIC_PATH = "__mini_css_extract_plugin_public_path_auto__";
const ABSOLUTE_PUBLIC_PATH = "webpack:///mini-css-extract-plugin/";
const BASE_URI = "webpack://";
const SINGLE_DOT_PATH_SEGMENT = "__mini_css_extract_plugin_single_dot_path_segment__";

/**
 * @param {string} str
 * @returns {boolean}
 */
function isAbsolutePath(str) {
  return path.posix.isAbsolute(str) || path.win32.isAbsolute(str);
}
const RELATIVE_PATH_REGEXP = /^\.\.?[/\\]/;

/**
 * @param {string} str
 * @returns {boolean}
 */
function isRelativePath(str) {
  return RELATIVE_PATH_REGEXP.test(str);
}

// TODO simplify for the next major release
/**
 * @param {LoaderContext} loaderContext
 * @param {string} request
 * @returns {string}
 */
function stringifyRequest(loaderContext, request) {
  if (typeof loaderContext.utils !== "undefined" && typeof loaderContext.utils.contextify === "function") {
    return JSON.stringify(loaderContext.utils.contextify(loaderContext.context || loaderContext.rootContext, request));
  }
  const splitted = request.split("!");
  const {
    context
  } = loaderContext;
  return JSON.stringify(splitted.map(part => {
    // First, separate singlePath from query, because the query might contain paths again
    const splittedPart = part.match(/^(.*?)(\?.*)/);
    const query = splittedPart ? splittedPart[2] : "";
    let singlePath = splittedPart ? splittedPart[1] : part;
    if (isAbsolutePath(singlePath) && context) {
      singlePath = path.relative(context, singlePath);
      if (isAbsolutePath(singlePath)) {
        // If singlePath still matches an absolute path, singlePath was on a different drive than context.
        // In this case, we leave the path platform-specific without replacing any separators.
        // @see https://github.com/webpack/loader-utils/pull/14
        return singlePath + query;
      }
      if (isRelativePath(singlePath) === false) {
        // Ensure that the relative path starts at least with ./ otherwise it would be a request into the modules directory (like node_modules).
        singlePath = `./${singlePath}`;
      }
    }
    return singlePath.replace(/\\/g, "/") + query;
  }).join("!"));
}

/**
 * @param {string} filename
 * @param {string} outputPath
 * @param {boolean} enforceRelative
 * @returns {string}
 */
function getUndoPath(filename, outputPath, enforceRelative) {
  let depth = -1;
  let append = "";

  // eslint-disable-next-line no-param-reassign
  outputPath = outputPath.replace(/[\\/]$/, "");
  for (const part of filename.split(/[/\\]+/)) {
    if (part === "..") {
      if (depth > -1) {
        // eslint-disable-next-line no-plusplus
        depth--;
      } else {
        const i = outputPath.lastIndexOf("/");
        const j = outputPath.lastIndexOf("\\");
        const pos = i < 0 ? j : j < 0 ? i : Math.max(i, j);
        if (pos < 0) {
          return `${outputPath}/`;
        }
        append = `${outputPath.slice(pos + 1)}/${append}`;

        // eslint-disable-next-line no-param-reassign
        outputPath = outputPath.slice(0, pos);
      }
    } else if (part !== ".") {
      // eslint-disable-next-line no-plusplus
      depth++;
    }
  }
  return depth > 0 ? `${"../".repeat(depth)}${append}` : enforceRelative ? `./${append}` : append;
}

/**
 *
 * @param {string | function} value
 * @returns {string}
 */
function stringifyLocal(value) {
  return typeof value === "function" ? value.toString() : JSON.stringify(value);
}
module.exports = {
  trueFn,
  findModuleById,
  evalModuleCode,
  compareModulesByIdentifier,
  MODULE_TYPE,
  AUTO_PUBLIC_PATH,
  ABSOLUTE_PUBLIC_PATH,
  BASE_URI,
  SINGLE_DOT_PATH_SEGMENT,
  stringifyRequest,
  stringifyLocal,
  getUndoPath
};