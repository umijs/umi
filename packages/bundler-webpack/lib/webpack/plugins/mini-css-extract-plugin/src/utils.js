"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findModuleById = findModuleById;
exports.evalModuleCode = evalModuleCode;
exports.compareModulesByIdentifier = compareModulesByIdentifier;
exports.MODULE_TYPE = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _module() {
  const data = _interopRequireDefault(require("module"));

  _module = function _module() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }

const MODULE_TYPE = 'css/mini-extract';
exports.MODULE_TYPE = MODULE_TYPE;

function findModuleById(compilation, id) {
  const modules = compilation.modules,
        chunkGraph = compilation.chunkGraph;

  var _iterator = _createForOfIteratorHelper(modules),
      _step;

  try {
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      const module = _step.value;
      const moduleId = typeof chunkGraph !== 'undefined' ? chunkGraph.getModuleId(module) : module.id;

      if (moduleId === id) {
        return module;
      }
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }

  return null;
}

function evalModuleCode(loaderContext, code, filename) {
  const module = new (_module().default)(filename, loaderContext);
  module.paths = _module().default._nodeModulePaths(loaderContext.context); // eslint-disable-line no-underscore-dangle

  module.filename = filename;

  module._compile(code, filename); // eslint-disable-line no-underscore-dangle


  return module.exports;
}

function compareIds(a, b) {
  if (typeof a !== typeof b) {
    return typeof a < typeof b ? -1 : 1;
  }

  if (a < b) {
    return -1;
  }

  if (a > b) {
    return 1;
  }

  return 0;
}

function compareModulesByIdentifier(a, b) {
  return compareIds(a.identifier(), b.identifier());
}