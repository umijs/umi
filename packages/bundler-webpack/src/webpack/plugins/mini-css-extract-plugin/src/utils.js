import NativeModule from 'module';

const MODULE_TYPE = 'css/mini-extract';

function findModuleById(compilation, id) {
  const { modules, chunkGraph } = compilation;

  for (const module of modules) {
    const moduleId =
      typeof chunkGraph !== 'undefined'
        ? chunkGraph.getModuleId(module)
        : module.id;

    if (moduleId === id) {
      return module;
    }
  }

  return null;
}

function evalModuleCode(loaderContext, code, filename) {
  const module = new NativeModule(filename, loaderContext);

  module.paths = NativeModule._nodeModulePaths(loaderContext.context); // eslint-disable-line no-underscore-dangle
  module.filename = filename;
  module._compile(code, filename); // eslint-disable-line no-underscore-dangle

  return module.exports;
}

function compareIds(a, b) {
  if (typeof a !== typeof b) {
    return typeof a < typeof b ? -1 : 1;
  }

  if (a < b) {
    return -1;
  }

  if (a > b) {
    return 1;
  }

  return 0;
}

function compareModulesByIdentifier(a, b) {
  return compareIds(a.identifier(), b.identifier());
}

export {
  MODULE_TYPE,
  findModuleById,
  evalModuleCode,
  compareModulesByIdentifier,
};
