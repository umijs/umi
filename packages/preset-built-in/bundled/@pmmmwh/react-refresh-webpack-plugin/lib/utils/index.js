const getAdditionalEntries = require('./getAdditionalEntries');
const getIntegrationEntry = require('./getIntegrationEntry');
const getRefreshGlobal = require('./getRefreshGlobal');
const getSocketIntegration = require('./getSocketIntegration');
const injectRefreshEntry = require('./injectRefreshEntry');
const injectRefreshLoader = require('./injectRefreshLoader');
const makeRefreshRuntimeModule = require('./makeRefreshRuntimeModule');
const normalizeOptions = require('./normalizeOptions');

module.exports = {
  getAdditionalEntries,
  getIntegrationEntry,
  getRefreshGlobal,
  getSocketIntegration,
  injectRefreshEntry,
  injectRefreshLoader,
  makeRefreshRuntimeModule,
  normalizeOptions,
};
