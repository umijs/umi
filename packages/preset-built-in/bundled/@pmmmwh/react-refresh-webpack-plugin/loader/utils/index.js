const getIdentitySourceMap = require('./getIdentitySourceMap');
const getModuleSystem = require('./getModuleSystem');
const getRefreshModuleRuntime = require('./getRefreshModuleRuntime');
const normalizeOptions = require('./normalizeOptions');

module.exports = {
  getIdentitySourceMap,
  getModuleSystem,
  getRefreshModuleRuntime,
  normalizeOptions,
};
