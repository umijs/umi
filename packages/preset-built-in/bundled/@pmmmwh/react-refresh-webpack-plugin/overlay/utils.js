/**
 * Debounce a function to delay invoking until wait (ms) have elapsed since the last invocation.
 * @param {function(...*): *} fn The function to be debounced.
 * @param {number} wait Milliseconds to wait before invoking again.
 * @return {function(...*): void} The debounced function.
 */
export function debounce(fn, wait) {
  /**
   * A cached setTimeout handler.
   * @type {number | undefined}
   */
  let timer;

  /**
   * @returns {void}
   */
  function debounced() {
    const context = this;
    const args = arguments;

    clearTimeout(timer);
    timer = setTimeout(function () {
      return fn.apply(context, args);
    }, wait);
  }

  return debounced;
}

/**
 * Prettify a filename from error stacks into the desired format.
 * @param {string} filename The filename to be formatted.
 * @returns {string} The formatted filename.
 */
export function formatFilename(filename) {
  // Strip away protocol and domain for compiled files
  const htmlMatch = /^https?:\/\/(.*)\/(.*)/.exec(filename);
  if (htmlMatch && htmlMatch[1] && htmlMatch[2]) {
    return htmlMatch[2];
  }

  // Strip everything before the first directory for source files
  const sourceMatch = /\/.*?([^./]+[/|\\].*)$/.exec(filename);
  if (sourceMatch && sourceMatch[1]) {
    return sourceMatch[1].replace(/\?$/, '');
  }

  // Unknown filename type, use it as is
  return filename;
}

/**
 * Remove all children of an element.
 * @param {HTMLElement} element A valid HTML element.
 * @param {number} [skip] Number of elements to skip removing.
 * @returns {void}
 */
export function removeAllChildren(element, skip) {
  /** @type {Node[]} */
  const childList = Array.prototype.slice.call(
    element.childNodes,
    typeof skip !== 'undefined' ? skip : 0
  );

  for (let i = 0; i < childList.length; i += 1) {
    element.removeChild(childList[i]);
  }
}
