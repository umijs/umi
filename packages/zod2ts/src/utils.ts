import type { ZodTypeAny } from '@umijs/utils/compiled/zod';
import ts from 'typescript';
import type { GetType, GetTypeFunction } from './types';
const { factory: f } = ts;

export const maybeIdentifierToTypeReference = (
  identifier: ts.Identifier | ts.TypeNode,
) => {
  if (ts.isIdentifier(identifier)) {
    return f.createTypeReferenceNode(identifier);
  }

  return identifier;
};

export const createTypeReferenceFromString = (identifier: string) =>
  f.createTypeReferenceNode(f.createIdentifier(identifier));

export const createUnknownKeywordNode = () =>
  f.createKeywordTypeNode(ts.SyntaxKind.UnknownKeyword);

export const createTypeAlias = (
  node: ts.TypeNode,
  identifier: string,
  comment?: string,
) => {
  const typeAlias = f.createTypeAliasDeclaration(
    undefined,
    f.createIdentifier(identifier),
    undefined,
    node,
  );

  if (comment) {
    addJsDocComment(typeAlias, comment);
  }

  return typeAlias;
};

export const printNode = (
  node: ts.Node,
  printerOptions?: ts.PrinterOptions,
) => {
  const sourceFile = ts.createSourceFile(
    'print.ts',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const printer = ts.createPrinter(printerOptions);
  return printer.printNode(ts.EmitHint.Unspecified, node, sourceFile);
};

export const withGetType = <T extends ZodTypeAny & GetType>(
  schema: T,
  getType: GetTypeFunction,
): T => {
  schema.getType = getType;
  return schema;
};

const identifierRE = /^[A-Za-z_$][A-Za-z0-9_$]*$/;

export const getIdentifierOrStringLiteral = (str: string) => {
  if (identifierRE.test(str)) {
    return f.createIdentifier(str);
  }

  return f.createStringLiteral(str);
};

export const addJsDocComment = (node: ts.Node, text: string) => {
  ts.addSyntheticLeadingComment(
    node,
    ts.SyntaxKind.MultiLineCommentTrivia,
    `* ${text} `,
    true,
  );
};
