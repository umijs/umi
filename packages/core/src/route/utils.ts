import { winPath } from '@umijs/utils';
import { extname } from 'path';

export function createRouteId(file: string) {
  return winPath(stripFileExtension(file));
}

export function stripFileExtension(file: string) {
  return file.replace(/\.[a-z0-9]+$/i, '');
}

export function byLongestFirst(a: string, b: string): number {
  return b.length - a.length;
}

const routeModuleExts = ['.js', '.jsx', '.ts', '.tsx', '.md', '.mdx', '.vue'];
export function isRouteModuleFile(opts: { file: string; exclude?: RegExp[] }) {
  // TODO: add cache strategy
  for (const excludeRegExp of opts.exclude || []) {
    if (
      opts.file &&
      excludeRegExp instanceof RegExp &&
      excludeRegExp.test(opts.file)
    ) {
      return false;
    }
  }
  return routeModuleExts.includes(extname(opts.file));
}
