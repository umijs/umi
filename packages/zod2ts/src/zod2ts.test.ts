import { zod as z } from '@umijs/utils';
import ts from 'typescript';
import { printNode, withGetType, zodToTs } from '.';

export const printNodeTest = (node: ts.Node) =>
  printNode(node, { newLine: ts.NewLineKind.LineFeed });

test('zod default', () => {
  enum Fruits {
    Apple = 'apple',
    Banana = 'banana',
    Cantaloupe = 'cantaloupe',
    A = 5,
  }

  const example2 = z.object({
    a: z.string(),
    b: z.number(),
    c: z.array(z.string()).nonempty().length(10),
    d: z.object({
      e: z.string(),
    }),
  });

  const pickedSchema = example2.partial();

  const nativeEnum = withGetType(z.nativeEnum(Fruits), (ts, _, options) => {
    const identifier = ts.factory.createIdentifier('Fruits');

    if (options.resolveNativeEnums) return identifier;

    return ts.factory.createTypeReferenceNode(identifier, undefined);
  });

  type ELazy = {
    a: string;
    b: ELazy;
  };

  const eLazy: z.ZodSchema<ELazy> = withGetType(
    z.lazy(() => e3),
    (ts, identifier) =>
      ts.factory.createIndexedAccessTypeNode(
        ts.factory.createTypeReferenceNode(
          ts.factory.createIdentifier(identifier),
          undefined,
        ),
        ts.factory.createLiteralTypeNode(ts.factory.createStringLiteral('b')),
      ),
  );

  const e3 = z.object({
    a: z.string(),
    b: eLazy,
  });

  const dateType = withGetType(z.instanceof(Date), (ts) =>
    ts.factory.createIdentifier('Date'),
  );

  const example = z.object({
    a: z.string(),
    b: z.number(),
    c: z.array(
      z.object({
        a: z.string(),
      }),
    ),
    d: z.boolean(),
    e: eLazy,
    f: z.union([z.object({ a: z.number() }), z.literal('hi')]),
    g: z.enum(['hi', 'bye']),
    h: z
      .number()
      .and(z.bigint())
      .and(z.number().and(z.string()))
      .transform((arg) => console.log(arg)),
    i: z.date(),
    j: z.undefined(),
    k: z.null(),
    l: z.void(),
    m: z.any(),
    n: z.unknown(),
    o: z.never(),
    p: z.optional(z.string()),
    q: z.nullable(pickedSchema),
    r: z.tuple([z.string(), z.number(), z.object({ name: z.string() })]),
    s: z.record(
      z.object({
        de: z.object({
          me: z
            .union([
              z.tuple([z.string(), z.object({ a: z.string() })]),
              z.bigint(),
            ])
            .array(),
        }),
      }),
    ),
    t: z.map(z.string(), z.array(z.object({ p: z.string() }))),
    u: z.set(z.string()),
    v: z.intersection(z.string(), z.number()).or(z.bigint()),
    w: z.promise(z.number()),
    x: z
      .function()
      .args(z.string().nullish().default('heo'), z.boolean(), z.boolean())
      .returns(z.string()),
    y: z.string().optional().default('hi'),
    z: z
      .string()
      .refine((val) => val.length > 10)
      .or(z.number())
      .and(z.bigint().nullish().default(1000n)),
    aa: nativeEnum,
    bb: dateType,
    cc: z.lazy(() => z.string()),
    dd: z.nativeEnum(Fruits),
    ee: z.discriminatedUnion('kind', [
      z.object({ kind: z.literal('circle'), radius: z.number() }),
      z.object({ kind: z.literal('square'), x: z.number() }),
      z.object({ kind: z.literal('triangle'), x: z.number(), y: z.number() }),
    ]),
  });

  const { node } = zodToTs(example, 'Example', { resolveNativeEnums: true });

  expect(printNodeTest(node)).toMatchSnapshot();
});

test('zod.discriminatedUnion()', () => {
  const ShapeSchema = z.discriminatedUnion('kind', [
    z.object({ kind: z.literal('circle'), radius: z.number() }),
    z.object({ kind: z.literal('square'), x: z.number() }),
    z.object({ kind: z.literal('triangle'), x: z.number(), y: z.number() }),
  ]);

  const { node } = zodToTs(ShapeSchema, 'Shape');

  expect(printNodeTest(node)).toMatchSnapshot();
});

test('zod.object()', () => {
  const userSchema = z.object({
    username: z.string(),
    age: z.number(),
    isAdmin: z.boolean(),
    createdAt: z.date(),
    undef: z.undefined(),
    nu: z.null(),
    vo: z.void(),
    an: z.any(),
    unknow: z.unknown(),
    nev: z.never(),
  });

  const { node } = zodToTs(userSchema, 'User');

  expect(printNodeTest(node)).toMatchSnapshot();
});

test('zod.describe()', () => {
  const schema = z.object({
    name: z.string().describe('The name of the item'),
    price: z.number().describe('The price of the item'),
  });

  const { node } = zodToTs(schema);

  expect(printNodeTest(node)).toMatchSnapshot();
});

test('zod.optional()', () => {
  const OptionalStringSchema = z.string().optional();
  const schema = z.object({
    optional: OptionalStringSchema,
    required: z.string(),
    transform: z
      .number()
      .optional()
      .transform((arg) => arg),
    or: z.number().optional().or(z.string()),
    tuple: z
      .tuple([
        z.string().optional(),
        z.number(),
        z.object({
          optional: z.string().optional(),
          required: z.string(),
        }),
      ])
      .optional(),
  });

  const { node } = zodToTs(schema);

  expect(printNodeTest(node)).toMatchSnapshot();
});

test('zod.nullable()', () => {
  const schema = z.object({
    username: z.string().nullable(),
  });

  const { node } = zodToTs(schema);

  expect(printNodeTest(node)).toMatchSnapshot();
});

test('zod enum', () => {
  enum Fruit {
    Apple = 'apple',
    Banana = 'banana',
    Cantaloupe = 'cantaloupe',
  }

  enum Color {
    Red,
    Green,
    Blue,
  }

  const schema = z.object({
    fruit: withGetType(z.nativeEnum(Fruit), (ts) =>
      ts.factory.createIdentifier('Fruit'),
    ),
    color: withGetType(z.nativeEnum(Color), (ts) =>
      ts.factory.createIdentifier('Color'),
    ),
  });

  const { node, store } = zodToTs(schema);

  const enums = store.nativeEnums
    .map((item) => {
      return printNode(item);
    })
    .join('\n');

  expect(enums).toMatchSnapshot();
  expect(printNodeTest(node)).toMatchSnapshot();
});

test('zod.function()', () => {
  const schema = z
    .function()
    .args(z.string().nullish().default('name'), z.boolean(), z.boolean())
    .returns(z.string());
  const { node } = zodToTs(schema, 'Function');
  expect(printNodeTest(node)).toMatchSnapshot();
});

test('zod.function() 2', () => {
  const schema = z.object({
    h: z
      .function()
      .args(
        z.object({ name: z.string(), price: z.number(), comment: z.string() }),
      )
      .describe('create an item'),
  });

  const { node } = zodToTs(schema);
  expect(printNodeTest(node)).toMatchSnapshot();
});
