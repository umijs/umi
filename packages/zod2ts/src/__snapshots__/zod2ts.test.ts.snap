// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zod default 1`] = `
"{
    a: string;
    b: number;
    c: {
        a: string;
    }[];
    d: boolean;
    e: Example["b"];
    f: {
        a: number;
    } | "hi";
    g: "hi" | "bye";
    h: (number & bigint) & (number & string);
    i: Date;
    j?: undefined;
    k: null;
    l?: void | undefined;
    m?: any;
    n?: unknown;
    o: never;
    p?: string | undefined;
    q: {
        a?: string | undefined;
        b?: number | undefined;
        c?: string[] | undefined;
        d?: {
            e: string;
        } | undefined;
    } | null;
    r: [
        string,
        number,
        {
            name: string;
        }
    ];
    s: {
        [x: string]: {
            de: {
                me: ([
                    string,
                    {
                        a: string;
                    }
                ] | bigint)[];
            };
        };
    };
    t: Map<string, {
        p: string;
    }[]>;
    u: Set<string>;
    v: (string & number) | bigint;
    w: Promise<number>;
    x: (args_0: (string | null), args_1: boolean, args_2: boolean, ...args_3: unknown[]) => string;
    y?: string;
    z: (string | number) & ((bigint | null));
    aa: Fruits;
    bb: Date;
    cc: Example;
    dd: unknown;
    ee: {
        kind: "circle";
        radius: number;
    } | {
        kind: "square";
        x: number;
    } | {
        kind: "triangle";
        x: number;
        y: number;
    };
}"
`;

exports[`zod enum 1`] = `
"enum Fruit {
    Apple = "apple",
    Banana = "banana",
    Cantaloupe = "cantaloupe"
}
enum Color {
    "0" = "Red",
    "1" = "Green",
    "2" = "Blue",
    Red = 0,
    Green = 1,
    Blue = 2
}"
`;

exports[`zod enum 2`] = `
"{
    fruit: Fruit;
    color: Color;
}"
`;

exports[`zod.describe() 1`] = `
"{
    /** The name of the item */
    name: string;
    /** The price of the item */
    price: number;
}"
`;

exports[`zod.discriminatedUnion() 1`] = `
"{
    kind: "circle";
    radius: number;
} | {
    kind: "square";
    x: number;
} | {
    kind: "triangle";
    x: number;
    y: number;
}"
`;

exports[`zod.function() 1`] = `"(args_0: (string | null), args_1: boolean, args_2: boolean, ...args_3: unknown[]) => string"`;

exports[`zod.function() 2 1`] = `
"{
    /** create an item */
    h: (args_0: {
        name: string;
        price: number;
        comment: string;
    }, ...args_1: unknown[]) => unknown;
}"
`;

exports[`zod.nullable() 1`] = `
"{
    username: string | null;
}"
`;

exports[`zod.object() 1`] = `
"{
    username: string;
    age: number;
    isAdmin: boolean;
    createdAt: Date;
    undef?: undefined;
    nu: null;
    vo?: void | undefined;
    an?: any;
    unknow?: unknown;
    nev: never;
}"
`;

exports[`zod.optional() 1`] = `
"{
    optional?: string | undefined;
    required: string;
    transform?: number | undefined;
    or?: (number | undefined) | string;
    tuple?: [
        string | undefined,
        number,
        {
            optional?: string | undefined;
            required: string;
        }
    ] | undefined;
}"
`;
